tests/unit_tests/verRctNonSemanticsSimple.cpp

```diff
@@ -0,0 +1,434 @@
+        }
+    }
+
+    return true;
+}
+
+/**
+ * @brief Perform expand_transaction_1 and Blockchain::expand_transaction_2 on a certain transaction
+ */
+static void expand_transaction_fully(cryptonote::transaction& tx, const rct::ctkeyM& input_pubkeys)
+{
+    const crypto::hash tx_prefix_hash = cryptonote::get_transaction_prefix_hash(tx);
+    CHECK_AND_ASSERT_THROW_MES(cryptonote::expand_transaction_1(tx, false), "expand 1 failed");
+    CHECK_AND_ASSERT_THROW_MES
+    (
+        cryptonote::Blockchain::expand_transaction_2(tx, tx_prefix_hash, input_pubkeys),
+        "expand 2 failed"
+    );
+    CHECK_AND_ASSERT_THROW_MES(!memcmp(&tx_prefix_hash, &tx.rct_signatures.message, 32), "message check failed");
+    CHECK_AND_ASSERT_THROW_MES(input_pubkeys == tx.rct_signatures.mixRing, "mixring check failed");
+    CHECK_AND_ASSERT_THROW_MES(check_tx_is_expanded(tx, input_pubkeys), "tx expansion check 2 failed");
+}
+
+/**
+ * @brief Mostly construct transaction from binary file and provided mix ring pubkeys
+ *
+ * Most important to us, this should populate the .rct_signatures.message and
+ * .rct_signatures.mixRings fields of the transaction.
+ *
+ * @param file_name relative file path in unit test data directory
+ * @param input_pubkeys manually retrived input pubkey destination / masks for each ring
+ * @return cryptonote::transaction the expanded transaction
+ */
+static cryptonote::transaction expand_transaction_from_bin_file_and_pubkeys
+(
+    const char* file_name,
+    const rct::ctkeyM& input_pubkeys
+)
+{
+    cryptonote::transaction transaction;
+
+    const boost::filesystem::path tx_json_path = unit_test::data_dir / file_name;
+    std::string tx_blob;
+    CHECK_AND_ASSERT_THROW_MES
+    (
+        epee::file_io_utils::load_file_to_string(tx_json_path.string(), tx_blob),
+        "loading file to string failed"
+    );
+
+    CHECK_AND_ASSERT_THROW_MES
+    (
+        cryptonote::parse_and_validate_tx_from_blob(tx_blob, transaction),
+        "TX blob could not be parsed"
+    );
+
+    expand_transaction_fully(transaction, input_pubkeys);
+
+    return transaction;
+}
+
+/**
+ * @brief Return whether a modification changes blob resulting from do_serialize()
+ */
+template <typename T>
+static bool modification_changes_do_serialize
+(
+    const T& og_obj,
+    std::function<void(T&)> obj_modifier_func,
+    bool expected_change
+)
+{
+    T modded_obj = og_obj;
+    obj_modifier_func(modded_obj);
+    const std::string og_blob = stringify_with_do_serialize(og_obj);
+    const std::string modded_blob = stringify_with_do_serialize(modded_obj);
+    const bool did_change = modded_blob != og_blob;
+    if (did_change != expected_change)
+    {
+        const std::string og_hex = epee::to_hex::string(epee::strspan<uint8_t>(og_blob));
+        const std::string modded_hex = epee::to_hex::string(epee::strspan<uint8_t>(modded_blob));
+        MERROR("unexpected: modded_blob '" << modded_hex << "' vs og_blob ' << " << og_hex << "'");
```

Comment:
Could make use of `AssertionResult`, `AssertionSuccess`, and `AssertionFailure` here, e.g.:
`
static testing::AssertionResult modification_changes_do_serialize
(
... og_obj,
... obj_modifier_func
// note removal of expected_change
)
{
  ..
  const std::string og_hex = epee::to_hex::string(epee::strspan<uint8_t>(og_blob));
  const std::string modded_hex = epee::to_hex::string(epee::strspan<uint8_t>(modded_blob));
  if (og_blob == modded_blob)
    return testing::AssertionFailure() <<  "serialized blobs match: modded_blob '" << modded_hex << "' og_blob '" << og_hex << "'";
  else
    return testing::AssertionSuccess() <<  "serialized blobs differ: modded_blob '" << modded_hex << "' og_blob '" << og_hex << "'"; 
```

Then, if you expect a change:
```
EXPECT_TRUE(modification_changes_do_serialize(og_obj, obj_modifier_func)
```
And if you expect no change:
EXPECT_FALSE(modification_changes_do_serialize(og_obj, obj_modifier_func)

Reply:
But sometimes I want the modification to *not* result in a serializable change. In that case, using `AssertionFailure` doesn't seem quite right in that case, since matching blobs is the success case.

Reply:
`AssertionFailure` and `Success` act like bools with a message attached (this exact use is very clearly documented in the gtest.h file), hence the call to `EXPECT_[False|True]`. I guess you can quibble with google's naming choices but that makes using the gtest library much more difficult.

Reply:
I suppose I should note that I think your premise misses the mark a bit:
> But sometimes I want the modification to not result in a serializable change. In that case, using AssertionFailure doesn't seem quite right in that case,

The assertion failing in this case is that "modification changes do serialize". So, the result from the function makes sense. Then the caller decides whether the scenario should or should not change do serialize, i.e. whether the function's result constitutes a test failure or success. Whereas, the current version tries to combine the two separable concerns, causing the boolean value to repeat in the function call and the assertion (i.e. `EXPECT_TRUE(..., true)` and `EXPECT_FALSE(..., false)`).